nemo/collections/tts/modules/transformer_dec24.py

```diff
@@ -0,0 +1,691 @@
+            if self.use_cache:
+                if self.cache['self_attn_output'] is not None:
+                    x_ = torch.cat([self.cache['self_attn_output'], x_], dim=1)
+                self.cache['self_attn_output'] = x_
+            x = x + self.layer_scale_self_attn(x_)
+            x = self.norm_self(x) * x_mask_inv_float
+
+        x_attn_prob = None
+        if self.has_xattn and cond is not None:
+            if self.layer_norm_method == 'pre':
+                x_normed = self.norm_xattn_query(x)
+                if self.use_cache and self.cache['memory'] is not None:
+                    memory = self.cache['memory']
+                else:
+                    memory = self.norm_xattn_memory(cond) if self.apply_norm_to_cond else cond
+                    if self.use_cache:
+                        self.cache['memory'] = memory
+
+                x_res, x_attn_prob = self.cross_attention(
+                    query=x_normed,
+                    query_mask=x_mask,
+                    memory=memory,
+                    memory_mask=cond_mask,
+                    attn_prior=attn_prior,
+                    dump_attention=dump_attention,
+                    idx=idx,
+                )
+                if self.use_cache:
+                    if self.cache['cross_attn_output'] is not None:
+                        x_res = torch.cat([self.cache['cross_attn_output'], x_res], dim=1)
+                    self.cache['cross_attn_output'] = x_res
+                x = x + self.layer_scale_cross_attn(x_res)  # unbounded
+                x = x * x_mask_inv_float
+            elif self.layer_norm_method == 'post':
+                x_res, x_attn_prob = self.cross_attention(
+                    query=x,
+                    query_mask=x_mask,
+                    memory=cond,
+                    memory_mask=cond_mask,
+                    attn_prior=attn_prior,
+                    dump_attention=dump_attention,
+                )
+                if self.use_cache:
+                    if self.cache['cross_attn_output'] is not None:
+                        x_res = torch.cat([self.cache['cross_attn_output'], x_res], dim=1)
+                    self.cache['cross_attn_output'] = x_res
+
+                x = (x + self.layer_scale_cross_attn(x_res)) * x_mask_inv_float
+                x = self.norm_xattn_query(x)
+
+        # mlp final projection
+        if self.layer_norm_method == 'pre':
+            x = x + self.layer_scale_ff(self.pos_ff(self.norm_pos_ff(x)))
+            x *= x_mask_inv_float
+        elif self.layer_norm_method == 'post':
+            x = x + self.layer_scale_ff(self.pos_ff(x))
+            x *= x_mask_inv_float
+            x = self.norm_pos_ff(x)
+
+        attn_probabilities = {'self_attn_probabilities': s_attn_prob, 'cross_attn_probabilities': x_attn_prob}
+
+        return {
+            'output': x,
+            'attn_probabilities': attn_probabilities,
+        }
+
+
+class TransformerStack(nn.Module):
```

Comment:
Let's rename this to Transformer instead of TransformerStack and add a docstring